File: 1396. Design Underground System/Solution.cpp
#include <string>
#include <unordered_map>
using namespace std;

class UndergroundSystem {
public:
    UndergroundSystem()
    {
    }

    void checkIn(int id, string stationName, int t)
    {
        memo[id] = make_pair(stationName, t);
    }

    void checkOut(int id, string stationName, int t)
    {
        pair<string, int>& p = memo[id];
        string key = p.first + " " + stationName;
        timer[key].first += t - p.second;
        timer[key].second++;
    }

    double getAverageTime(string startStation, string endStation)
    {
        string key = startStation + " " + endStation;
        return (double)timer[key].first / timer[key].second;
    }

private:
    unordered_map<int, pair<string, int>> memo;
    unordered_map<string, pair<int, int>> timer;
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */

File: 1396. Design Underground System/Solution.py
class UndergroundSystem:
    def __init__(self):
        self.memo = {}
        self.timer = {}

    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.memo[id] = (stationName, t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        startStation, start = self.memo[id]
        key = startStation + " " + stationName
        total, cnt = self.timer.setdefault(key, (0, 0))
        self.timer[key] = (total + t - start, cnt + 1)

    def getAverageTime(self, startStation: str, endStation: str) -> float:
        key = startStation + " " + endStation
        total, cnt = self.timer[key]
        return total / cnt


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)
